### ISSUE TITLE
False positive import/no-deprecated for typescript overloads

### ISSUE THREAD

**Issue by ccjmne** [import-js/eslint-plugin-import#1532]
Consider this external library I'm consuming:

```typescript
export declare function doSomethingWith(items: string[]): string;

/** @deprecated Pass arguments in a single array instead `doSomethingWith([a, b, c])` */
export declare function doSomethingWith(...items: string[]): string;
```

The `import/no-deprecated` rule currently flags even the proper usages:

```typescript
import { doSomethingWith } from 'external-library'; // flagged

doSomethingWith('a', 'b', 'c'); // flagged, rightly so
doSomethingWith(['a', 'b', 'c']); // flagged, wrongly so
```

TSLint's `deprecation` rule gets it right (not bashing ESLint here at all, just pointing to a working implementation).

---

Additionally, in the following case:
```typescript
export declare function doSomethingWith(items: string[]): string;

/** @deprecated Pass arguments in a single array instead `doSomethingWith([a, b, c])` */
export declare function doSomethingWith(...items: string[]): string;

/** @deprecated Please don't use this method anymore, it's actually gonna blow everything up */
export declare function doSomethingWith(unused: number, ...items: string): string;
```

The deprecation warning displayed is that of the last overloaded method (AFAICT), and it may not make sense.

```typescript
import { doSomethingWith } from 'external-library'; // "Please don't use this method anymore..."

doSomethingWith('a', 'b', 'c'); // "Please don't use this method anymore..."
doSomethingWith(['a', 'b', 'c']); // "Please don't use this method anymore..."
```

**Comment by digeomel** [comment]
I just stumbled across the same bug. We're using the RxJs `forkJoin` method, which used to allow multiple observable parameters, as such:

`forkJoin(a$, b$, c$)`

but now it deprecated in favor of accepting an array of observables:

`forkJoin([a$, b$, c$])`

So, we're using the latter and it's reported as an error by eslint.
Any updated on this?

**Comment by ljharb** [comment]
@digeomel no, nobody's submitted a PR yet.

**Comment by digeomel** [comment]
@ljharb that's unfortunate, because that's a breaking change for us, moving from tslint. I can see now why the Angular team is still stuck to tslint. Makes me think of switching back too.

**Comment by ljharb** [comment]
Considering it's over a year deprecated (and should have been an eslint parser the whole time anyways), I'd say it's worth making the switch regardless of how many breaks there are 

You're welcome to make a PR yourself :-)

**Comment by json-derulo** [comment]
I've stumbled upon this issue as well. Today I found the plugin [eslint-plugin-deprecation](https://github.com/gund/eslint-plugin-deprecation). And with this plugin, the detection of deprecated symbols seems to work reliably for me, also with TypeScript overloads.

**Comment by ljharb** [comment]
@JounQin I don't think this should be closed; we should be fixing the rule, and making it more efficient.

**Comment by JounQin** [comment]
> @JounQin I don't think this should be closed; we should be fixing the rule, and making it more efficient.

@ljharb

Seems no one is working on it for years, and this is specific to TypeScript only? Isn't that out of scope?

**Comment by ljharb** [comment]
It's not out of scope at all, and it doesn't matter how old something is, "staleness" is irrelevant :-) it'll get worked on eventually.

**Comment by markchagers** [comment]
I'm seeing this with the ngMocks library.
When I put this line at the top of a test suite:
```javascript
MockInstance.scope('case');
```
the line is flagged as deprecated. However, the [scope function](https://github.com/help-me-mom/ng-mocks/blob/ef5f40a6743f0c767fba9e86ce13a7deca5bf8d4/libs/ng-mocks/src/lib/mock-instance/mock-instance.ts#L282) is not deprecated at all, just [some other overload](https://github.com/help-me-mom/ng-mocks/blob/ef5f40a6743f0c767fba9e86ce13a7deca5bf8d4/libs/ng-mocks/src/lib/mock-instance/mock-instance.ts#L200) of the MockInstance function is. The deprecation text shown is the one belonging to the latter signature.

**Comment by anenger** [comment]
Just bumping this old issue :) I had the same problem with the `getFocusStyle` function found here: https://github.com/microsoft/fluentui/blob/master/packages/style-utilities/src/styles/getFocusStyle.ts

There is an object parameter version which is not deprecated, but eslint thinks it is because of this issue.

**Comment by leighman** [comment]
Seems like the rule just gets a `metadata` object with `tags` including `@deprecated` for a particular method. Would that be coming from `@typescript-eslint/parser`?

**Comment by pfeileon** [comment]
This happens for RxJS's `exhaustMap`, `map`, `merge`, `mergeMap`, `switchMap`, `tap`, `timer`, `throwError` and NgRx's `createSelector` functions, as well.
